reporter/src/main/kotlin/model/EvaluatedModel.kt

```diff
@@ -19,7 +19,284 @@
+            PathExclude::class.java,
+            RuleViolationResolution::class.java,
+            ScopeName::class.java,
+            ScopeExclude::class.java
+        )
+
+        private val MAPPER_CONFIG: ObjectMapper.() -> Unit = {
+            registerKotlinModule()
+
+            registerModule(JavaTimeModule())
+            registerModule(IntIdModule(INT_ID_TYPES))
+
+            disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS)
+
+            propertyNamingStrategy = PROPERTY_NAMING_STRATEGY
+        }
+
+        private val JSON_MAPPER = JsonMapper().apply { MAPPER_CONFIG() }
+
+        fun create(input: ReporterInput): EvaluatedModel = EvaluatedModelBuilder(input).build()
+    }
+
+    fun toJson(): String {
+        val tree = JSON_MAPPER.valueToTree<ObjectNode>(this)
+        tree.forEach { node ->
+            if (node is ArrayNode) {
+                if (!node.isEmpty && node[0].has("_id")) {
+                    val sortedChildren =
+                        node.elements().asSequence().toList().sortedBy { it["_id"].intValue() }.toList()
+                    node.removeAll()
+                    node.addAll(sortedChildren)
+                }
+            }
+        }
+        return tree.toPrettyString()
+    }
+}
+
+@JsonIgnoreProperties(value = ["paths"], allowGetters = true)
+data class EvaluatedPackage(
+    val id: Identifier,
+    val isProject: Boolean,
+    val definitionFilePath: String,
+    val purl: String = id.toPurl(),
+    val declaredLicenses: List<LicenseId>,
+    val declaredLicensesProcessed: EvaluatedProcessedDeclaredLicense,
+    val detectedLicenses: Set<LicenseId>,
+    @JsonInclude(JsonInclude.Include.NON_NULL)
+    val concludedLicense: SpdxExpression? = null,
+    val description: String,
+    val homepageUrl: String,
+    val binaryArtifact: RemoteArtifact,
+    val sourceArtifact: RemoteArtifact,
+    val vcs: VcsInfo,
+    val vcsProcessed: VcsInfo = vcs.normalize(),
+    val curations: List<PackageCurationResult>,
+    @JsonIdentityReference(alwaysAsId = true)
+    var paths: MutableList<EvaluatedPackagePath> = mutableListOf(),
+    val levels: SortedSet<Int>,
+    val scopes: MutableSet<ScopeName>,
+    val scanResults: List<EvaluatedScanResult>,
+    val findings: List<EvaluatedFinding>,
+    val isExcluded: Boolean,
+    val pathExcludes: List<PathExclude>,
+    val scopeExcludes: List<ScopeExclude>,
+    val issues: List<EvaluatedOrtIssue>
+)
+
+data class EvaluatedPackagePath(
+    val pkg: EvaluatedPackage,
+    val project: EvaluatedPackage,
+    val scope: ScopeName,
+    val path: List<EvaluatedPackage>
+)
+
+data class EvaluatedScanResult(
+    val provenance: Provenance,
+    val scanner: ScannerDetails,
+    val startTime: Instant,
+    val endTime: Instant,
+    val fileCount: Int,
+    val packageVerificationCode: String,
+    val issues: List<EvaluatedOrtIssue>
+)
+
+enum class EvaluatedFindingType {
+    COPYRIGHT, LICENSE
+}
+
+data class EvaluatedFinding(
```

Comment:
Does this class have the following assumptions?
- `type == LICENSE <--> license != null && copyright == null`
- `type == LICENSE <--> license == null && copyright != null`

Reply:
Yes.

Reply:
How about avoiding such assumptions by defining the data class as:
`EvaluatedFinding(startLine, endline, path, scanResult, value)` 
where value is either a `copyrightstatement` or a `licenseId` and hold the license and copyright findings in separate collections so they don't need to have that `type` field.

Maybe it's just my lack of understanding of the usecase, reason why you choose to model it this way. Can you give a hint?

Reply:
This comes from a requirement of the web-app-reporter which will show the findings in exactly the order they are added here. I assume this part will need changes after the first iteration of the rewritten web app reporter is implemented. It's also the reason I haven't documented many fields, because I first want to know if this model is stable or needs bigger changes before spending time writing documentation.